src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/halapi.cxx

```diff
@@ -100,32 +98,38 @@ xclDeviceHandle xclOpen(unsigned deviceIndex, const char *logfileName, xclVerbos
         std::cout << sDummyDeviceMsg << std::endl;
     }
   }
-  return (xclDeviceHandle *)handle;
+  return (xclDeviceHandle *)handle.get();
 }
 
 void xclClose(xclDeviceHandle handle)
 {
-  xclcpuemhal2::CpuemShim *drv = xclcpuemhal2::CpuemShim::handleCheck(handle);
+  auto drv = xclcpuemhal2::CpuemShim::handleCheck(handle);
   if (!drv)
     return ;
   drv->xclClose();
-  if (xclcpuemhal2::CpuemShim::handleCheck(handle) && xclcpuemhal2::devices.size() == 0) {
-    delete ((xclcpuemhal2::CpuemShim*)handle);
+  if (xclcpuemhal2::CpuemShim::handleCheck(handle)){ 
```

Comment:
I am not understanding this check ?  Isn't it the same check as on line 106, so if `drv` is valid we come here and make the same check again?
```suggestion
  if (xclcpuemhal2::CpuemShim::handleCheck(handle)) { 
```